### PR TITLE
feat(zk_toolbox): add test upgrade subcommand to zk_toolbox

### PR DESCRIPTION
**Pull request by manuelmauro**
## What ❔

Add upgrade tests to zk_supervisor

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/tests/upgrade-test/tests/tester.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ export class Tester {
             ethWallet = new ethers.Wallet(process.env.MASTER_WALLET_PK!);
         }
         ethWallet = ethWallet.connect(ethProvider);
-        const web3Provider = new zksync.Provider(process.env.ZKSYNC_WEB3_API_URL || process.env.API_WEB3_JSON_RPC_HTTP_URL ||  "http://localhost:3050");
+        const web3Provider = new zksync.Provider(process.env.ZKSYNC_WEB3_API_URL || process.env.API_WEB3_JSON_RPC_HTTP_URL || "http://localhost:3050");
```
Comment (Deniallugo) [inline core/tests/upgrade-test/tests/tester.ts:40]: please use data from the config.

File: core/tests/upgrade-test/tests/upgrade.test.ts [hunk 2]
```diff
@@ -7,40 +7,32 @@ import fs from 'fs';
...
-    require(`${L1_CONTRACTS_FOLDER}/governance/ChainAdmin.sol/ChainAdmin.json`).abi
-);
-const L2_FORCE_DEPLOY_UPGRADER_ABI = new ethers.Interface(
-    require(`${process.env.ZKSYNC_HOME}/contracts/l2-contracts/artifacts-zk/contracts/ForceDeployUpgrader.sol/ForceDeployUpgrader.json`).abi
-);
-const COMPLEX_UPGRADER_ABI = new ethers.Interface(
-    require(`${process.env.ZKSYNC_HOME}/contracts/system-contracts/artifacts-zk/contracts-preprocessed/ComplexUpgrader.sol/ComplexUpgrader.json`).abi
-);
-const COUNTER_BYTECODE =
-    require(`${process.env.ZKSYNC_HOME}/core/tests/ts-integration/artifacts-zk/contracts/counter/counter.sol/Counter.json`).deployedBytecode;
```
Comment (Deniallugo) [inline core/tests/upgrade-test/tests/upgrade.test.ts:31]: everywhere with `ZKSYNC_HOME` use `pathToHome = path.join(__dirname, '../../../..`

File: core/tests/upgrade-test/tests/upgrade.test.ts [hunk 3]
```diff
@@ -50,16 +42,56 @@ describe('Upgrade test', function () {
...
+        complexUpgraderAddress = '0x000000000000000000000000000000000000800f';
+
+        if (fileConfig.loadFromFile) {
+            const contractsConfig = loadConfig({ pathToHome, chain: fileConfig.chain, config: 'contracts.yaml' });
+            const secretsConfig = loadConfig({ pathToHome, chain: fileConfig.chain, config: 'secrets.yaml' });
+
+            ethProviderAddress = secretsConfig.l1.l1_rpc_url;
+            contractsL2DefaultUpgradeAddr = contractsConfig.l2.default_l2_upgrader;
+            process.env.CONTRACTS_DEFAULT_UPGRADE_ADDR = contractsConfig.l1.default_upgrade_addr;
+            process.env.CONTRACTS_PRIORITY_TX_MAX_GAS_LIMIT = '72000000';
```
Comment (Deniallugo) [inline core/tests/upgrade-test/tests/upgrade.test.ts:62]: Don't use envs inside tests.

File: core/tests/upgrade-test/tests/utils.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    chainAdminAbi: any;
+    l2ForceDeployUpgraderAbi: any;
+    complexUpgraderAbi: any;
+    counterBytecode: any;
+    stateTransitonManager: any;
+}
+
+export function initContracts(zkToolbox: boolean): Contracts {
+    if (zkToolbox) {
+        const CONTRACTS_FOLDER = `${process.env.ZKSYNC_HOME}/contracts`;
```
Comment (Deniallugo) [inline core/tests/upgrade-test/tests/utils.ts:69]: don't use ZKSYNC_HOME

File: zk_toolbox/crates/zk_inception/src/commands/contract/build.rs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+use crate::messages::MSG_CONTRACT_BUILDING;
+
+pub(crate) async fn run(shell: &Shell) -> anyhow::Result<()> {
+    let ecosystem = EcosystemConfig::from_file(shell)?;
+
+    let _dir_guard = shell.push_dir(&ecosystem.link_to_code);
+    let spinner = Spinner::new(MSG_CONTRACT_BUILDING);
+
+    Cmd::new(cmd!(shell, "yarn l1-contracts build")).run()?;
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/contract/build.rs:13]: we don't need to build l1 contracts we use forge for it

File: core/tests/upgrade-test/tests/utils.ts (hunk 4 above)
Comment (manuelmauro) [reply core/tests/upgrade-test/tests/utils.ts]: Addressed in https://github.com/matter-labs/zksync-era/pull/2515/commits/0f31c9ff8b170ebd32aa9fe32f99458a46610628

File: zk_toolbox/crates/zk_inception/src/commands/contract/build.rs (hunk 5 above)
Comment (manuelmauro) [reply zk_toolbox/crates/zk_inception/src/commands/contract/build.rs]: Removed in https://github.com/matter-labs/zksync-era/pull/2515/commits/d42c7604f48ba1633f2b1760d0d620c507cfd45b

File: core/tests/upgrade-test/tests/upgrade.test.ts (hunk 2 above)
Comment (manuelmauro) [reply core/tests/upgrade-test/tests/upgrade.test.ts:31]: Addressed in https://github.com/matter-labs/zksync-era/commit/0f31c9ff8b170ebd32aa9fe32f99458a46610628

File: core/tests/upgrade-test/tests/tester.ts (hunk 1 above)
Comment (manuelmauro) [reply core/tests/upgrade-test/tests/tester.ts]: Addressed in https://github.com/matter-labs/zksync-era/pull/2515/commits/315b7c813752d5733d8ee85d0706f133a7d8c5da

File: core/tests/upgrade-test/tests/upgrade.test.ts [hunk 6]
```diff
@@ -50,16 +42,68 @@ describe('Upgrade test', function () {
...
+
+            upgradeAddress = process.env.CONTRACTS_DEFAULT_UPGRADE_ADDR;
+            if (!upgradeAddress) {
+                throw new Error('CONTRACTS_DEFAULT_UPGRADE_ADDR not set');
+            }
+            contractsPriorityTxMaxGasLimit = process.env.CONTRACTS_PRIORITY_TX_MAX_GAS_LIMIT!;
+        }
+
+        const network = process.env.CHAIN_ETH_NETWORK || 'localhost';
+        tester = await Tester.init(network, ethProviderAddress!, web3JsonRpc!);
```
Comment (Deniallugo) [inline core/tests/upgrade-test/tests/upgrade.test.ts:81]: let's slightly change the approach with `network` don't pass this variable and just check the existence of `MASTER_WALLET_PK` inside tester and apply the corresponding configs

File: core/tests/upgrade-test/tests/upgrade.test.ts (hunk 3 above)
Comment (manuelmauro) [reply core/tests/upgrade-test/tests/upgrade.test.ts]: Addressed in https://github.com/matter-labs/zksync-era/pull/2515/commits/f7c285e6db84b64769a2d4453c1c00789152fa6c

File: core/tests/upgrade-test/tests/upgrade.test.ts (hunk 6 above)
Comment (manuelmauro) [reply core/tests/upgrade-test/tests/upgrade.test.ts]: Addressed in https://github.com/matter-labs/zksync-era/pull/2515/commits/9f0d1af6c01e433ccae43fc3965b0994ccbd8fc2

**Review by Deniallugo — APPROVED** [review]

**Review by sysadmin-of-your-crap — APPROVED** [review]

File: zk_toolbox/crates/zk_supervisor/src/messages.rs [hunk 7]
```diff
@@ -73,6 +73,7 @@ pub(super) const MSG_DATABASE_NEW_MIGRATION_SUCCESS: &str = "Migration created s
 pub(super) const MSG_INTEGRATION_TESTS_ABOUT: &str = "Run integration tests";
 pub(super) const MSG_REVERT_TEST_ABOUT: &str = "Run revert tests";
 pub(super) const MSG_RECOVERY_TEST_ABOUT: &str = "Run recovery tests";
+pub(super) const MSG_UPGRADE_TEST_ABOUT: &str = "Run recovery tests";
```
Comment (perekopskiy) [inline zk_toolbox/crates/zk_supervisor/src/messages.rs:76]: nit: change message

File: zk_toolbox/crates/zk_supervisor/src/messages.rs [hunk 8]
```diff
@@ -113,6 +114,12 @@ pub(super) const MSG_REVERT_TEST_RUN_SUCCESS: &str = "Revert and restart test ra
 pub(super) const MSG_RECOVERY_TEST_RUN_INFO: &str = "Running recovery test";
 pub(super) const MSG_RECOVERY_TEST_RUN_SUCCESS: &str = "Recovery test ran successfully";
 
+// Upgrade tests related messages
+pub(super) const MSG_UPGRADE_TEST_RUN_INFO: &str = "Running upgrade test";
+pub(super) const MSG_UPGRADE_TEST_INSTALLING_DEPENDENCIES: &str =
+    "Building and installing dependencies. This process may take a lot of time...";
+pub(super) const MSG_UPGRADE_TEST_RUN_SUCCESS: &str = "Revert and restart test ran successfully";
```
Comment (perekopskiy) [inline zk_toolbox/crates/zk_supervisor/src/messages.rs:121]: same

**Review by perekopskiy — APPROVED** [review]

File: zk_toolbox/crates/zk_supervisor/src/messages.rs (hunk 7 above)
Comment (manuelmauro) [reply zk_toolbox/crates/zk_supervisor/src/messages.rs]: Thank you! Addressed in https://github.com/matter-labs/zksync-era/pull/2515/commits/78799acf728d49383e25de636499e6948d57f72b

File: zk_toolbox/crates/zk_supervisor/src/messages.rs (hunk 8 above)
Comment (manuelmauro) [reply zk_toolbox/crates/zk_supervisor/src/messages.rs]: Addressed in https://github.com/matter-labs/zksync-era/pull/2515/commits/78799acf728d49383e25de636499e6948d57f72b

**Review by Deniallugo — APPROVED** [review]

**Review by onyxet — APPROVED** [review]

**Pull request merged**